### PR TITLE
8311821: Simplify ParallelGCThreadsConstraintFunc after CMS removal

### PR DESCRIPTION
**Pull request by kspeeyu**
Code oddities reported by SonarCloud after CMS removal
Fix: moving ParallelGCThreadsConstraintFuncParallel to shared code as ParallelGCThreadsConstraintFunc

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311821](https://bugs.openjdk.org/browse/JDK-8311821): Simplify ParallelGCThreadsConstraintFunc after CMS removal (**Enhancement** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14965/head:pull/14965` \
`$ git checkout pull/14965`

Update a local copy of the PR: \
`$ git checkout pull/14965` \
`$ git pull https://git.openjdk.org/jdk.git pull/14965/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14965`

View PR using the GUI difftool: \
`$ git pr show -t 14965`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14965.diff">https://git.openjdk.org/jdk/pull/14965.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14965#issuecomment-1644785981)

### CODE REVIEW

File: src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp [hunk 1]
```diff
@@ -53,18 +53,19 @@
...
 JVMFlag::Error ParallelGCThreadsConstraintFunc(uint value, bool verbose) {
-  JVMFlag::Error status = JVMFlag::SUCCESS;
+  // Parallel GC passes ParallelGCThreads when creating GrowableArray as 'int' type parameter.
+  // So can't exceed with "max_jint"
 
-#if INCLUDE_PARALLELGC
-  status = ParallelGCThreadsConstraintFuncParallel(value, verbose);
-  if (status != JVMFlag::SUCCESS) {
-    return status;
+  if (UseParallelGC && (value > (uint)max_jint)) {
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp:61]: Throughout this change: s/max_jint/INT_MAX/

File: src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp [hunk 2]
```diff
@@ -53,18 +53,19 @@
 // checking functions,  FLAG_IS_CMDLINE() is used to check if
 // the flag has been set by the user and so should be checked.
 
-// As ParallelGCThreads differs among GC modes, we need constraint function.
+// (earlier ParallelGCThreadConstraintsFuncParallel) replaced after CMS is removed
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp:56]: I don't think we need this comment.

File: src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp [hunk 3]
```diff
@@ -53,18 +53,19 @@
...
-#if INCLUDE_PARALLELGC
-  status = ParallelGCThreadsConstraintFuncParallel(value, verbose);
-  if (status != JVMFlag::SUCCESS) {
-    return status;
+  if (UseParallelGC && (value > (uint)max_jint)) {
+    JVMFlag::printError(verbose,
+                        "ParallelGCThreads (" UINT32_FORMAT ") must be "
+                        "less than or equal to " UINT32_FORMAT " for Parallel GC\n",
+                        value, max_jint);
+    return JVMFlag::VIOLATES_CONSTRAINT;
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/jvmFlagConstraintsGC.cpp:66]: There is no reason to use UINT32_FORMAT here, and indeed it just clutters things and makes it harder to read
and understand (and could even be wrong on some theoretical platform with ints larger than 32bits).
Just using "%u" for the first, and "%d" for the second is simple and correct.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
Could we eliminate the constraint function and instead specify a range for the option?
Using `range(0, INT_MAX)` would reduce the upper bound for collectors other than
ParallelGC, but I doubt anyone would notice :)

Or keep the constraint function if we can also move the check here that the value must be
greater than zero for most (i.e. not SerialGC) collectors.  (I think that's in ergo and might
not be movable because of other ergo calculations based on it that might not work for
a value of 0.)

**Comment by shipilev** [comment]
> Could we eliminate the constraint function and instead specify a range for the option? Using `range(0, INT_MAX)` would reduce the upper bound for collectors other than ParallelGC, but I doubt anyone would notice :)

I agree! Let's just put the `range(0, INT_MAX)` here: https://github.com/openjdk/jdk/blob/8cd43bff3cd18d6e83cbf07b78a809ad002993c5/src/hotspot/share/gc/shared/gc_globals.hpp#L136

**Comment by shipilev** [comment]
I don't think we need `ParallelGCThreadsConstraintFunc` at all now, seeing how we removed its only (?) use.

**Comment by kspeeyu** [comment]
> I don't think we need `ParallelGCThreadsConstraintFunc` at all now, seeing how we removed its only (?) use.

okay, so we remove 'ParallelGCThreadsConstraintFunc' as we replaced it with specifying a range instead? got it.

File: src/hotspot/share/gc/shared/gc_globals.hpp [hunk 4]
```diff
@@ -133,7 +133,7 @@
                                                                             \
   product(uint, ParallelGCThreads, 0,                                       \
           "Number of parallel threads parallel gc will use")                \
-          constraint(ParallelGCThreadsConstraintFunc,AfterErgo)             \
+          range(0, INT_MAX)                                                 \
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/gc_globals.hpp:136]: A comment to explain the range restriction seems useful.

**Review by kimbarrett — APPROVED** [review]
Other than the suggestion for an additional comment, this looks good.

File: src/hotspot/share/gc/shared/gc_globals.hpp (hunk 4 above)
Comment (shipilev) [reply src/hotspot/share/gc/shared/gc_globals.hpp:136]: +1. Something like:

```
  /* notice: the max range value here is INT_MAX not UINT_MAX */            \
  /* to protect from overflows                                */            \
  product(uint, ParallelGCThreads, 0,                                       \
          "Number of parallel threads parallel gc will use")                \
          range(0, INT_MAX)                                                 \
```

**Review by shipilev — APPROVED** [review]
Looks good, with minor nit about overflow comment.

**Comment by shipilev** [comment]
Ah, see, there is a failing test:
https://github.com/kspeeyu/jdk/actions/runs/5651211226#user-content-gc_arguments_testparallelgcthreads

```
----------System.err:(22/1299)----------
 stdout: [];
 stderr: [uint ParallelGCThreads=4294967295 is outside the allowed range [ 0 ... 2147483647 ]
Improperly specified VM option 'ParallelGCThreads=4294967295'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
]
 exitValue = 1
```

We need to adjust that test for new boundary now, I think.

File: test/hotspot/jtreg/gc/arguments/TestParallelGCThreads.java [hunk 5]
```diff
@@ -113,14 +113,14 @@ public static void testFlags() throws Exception {
       }
     }
 
-    // 4294967295 == (unsigned int) -1
-    // So setting ParallelGCThreads=4294967295 should give back 4294967295
+    // 2147483647 == (signed int) -1
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/arguments/TestParallelGCThreads.java:116]: This is not technically correct: `(signed int) -1` is `-1`.
Rewrite the whole comment to: `// Test the max value for ParallelGCThreads`.

**Review by shipilev — APPROVED** [review]
This looks okay. @kimbarrett, you good with modifying the test a little?

**Comment by shipilev** [comment]
Or maybe @tschatzl ?

**Review by kimbarrett — APPROVED** [review]
Still looks good.

**Review by tschatzl — APPROVED** [review]

**Pull request closed without merging**